File: src/main/java/io/outfoxx/typescriptpoet/FileSpec.kt
/*
 * Copyright 2017 Outfox, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.outfoxx.typescriptpoet

import java.io.File
import java.io.IOException
import java.io.OutputStreamWriter
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths


/**
 * A TypeScript file containing top level objects like classes, objects, functions, properties, and type
 * aliases.
 *
 * Items are output in the following order:
 * - Comment
 * - Imports
 * - Members
 */
class FileSpec
private constructor(
   builder: Builder
) {
  val path = builder.path
  val comment = builder.comment.build()
  val members = builder.members.toList()
  val indent = builder.indent


  fun exportType(typeName: String): TypeName? {
    val typeNameParts = typeName.split('.')
    return exportNamed(typeNameParts.first())?.let {
      TypeName.anyType(typeName, it)
    }
  }

  fun exportNamed(symbolName: String): SymbolSpec? {
    return members
       .map {
         when (it) {
           is ModuleSpec -> it.name
           is ClassSpec -> it.name
           is InterfaceSpec -> it.name
           is EnumSpec -> it.name
           is FunctionSpec -> it.name
           is PropertySpec -> it.name
           is TypeAliasSpec -> it.name
           else -> throw IllegalStateException("unrecognized member type")
         }
       }
       .firstOrNull { it == symbolName }
       ?.let { SymbolSpec.importsName(symbolName, "!" + path) }
  }

  fun exportAll(localName: String): SymbolSpec {
    return SymbolSpec.importsAll(localName, "!" + path)
  }

  @Throws(IOException::class)
  fun writeTo(out: Appendable) {
    // First pass: emit the entire class, just to collect the types we'll need to import.
    val importsCollector = CodeWriter(NullAppendable, indent)
    emit(importsCollector)
    val requiredImports = importsCollector.requiredImports()

    // Second pass: write the code, taking advantage of the imports.
    val codeWriter = CodeWriter(out, indent, requiredImports)
    emit(codeWriter)
  }

  /** Writes this to `directory` as UTF-8 using the standard directory structure.  */
  @Throws(IOException::class)
  fun writeTo(directory: Path) {
    require(Files.notExists(directory) || Files.isDirectory(directory)) {
      "path $directory exists but is not a directory."
    }
    val outputPath = directory.resolve("$path.ts")
    OutputStreamWriter(Files.newOutputStream(outputPath), UTF_8).use { writer -> writeTo(writer) }
  }

  /** Writes this to `directory` as UTF-8 using the standard directory structure.  */
  @Throws(IOException::class)
  fun writeTo(directory: File) = writeTo(directory.toPath())

  private fun emit(codeWriter: CodeWriter) {
    if (comment.isNotEmpty()) {
      codeWriter.emitComment(comment)
    }

    val imports = codeWriter.requiredImports()
    val augmentImports = codeWriter.requiredImports()
       .filterIsInstance<SymbolSpec.Augmented>()
       .groupBy { it.augmented }

    if (imports.isNotEmpty()) {
      imports
         .groupBy { FileModules.importPath(path, it.source) }
         .toSortedMap()
         .forEach { (sourceImportPath, imports) ->
           if (path == sourceImportPath || Paths.get(".").resolve(path) == sourceImportPath) {
             return@forEach
           }

           imports.filterIsInstance<SymbolSpec.ImportsAll>().forEach { import ->
             // Output star imports individually
             codeWriter.emitCode("%[import * as %L from '%L';\n%]", import.value, sourceImportPath)
             // Output related augments
             augmentImports[import.value]?.forEach { augment ->
               codeWriter.emitCode("%[import '%L';\n%]", augment.source)
             }
           }

           imports.filterIsInstance<SymbolSpec.ImportsName>()
              .map { it.value }
              .toSortedSet()
              .let { names ->
                if (names.isEmpty()) return@let
                // Output named imports as a group
                codeWriter
                   .emitCode("import {")
                   .indent()
                   .emitCode(names.joinToString(", "))
                   .unindent()
                   .emitCode("} from '%L';\n", sourceImportPath)
                // Output related augments
                names.forEach { name ->
                  augmentImports[name]?.forEach { augment ->
                    codeWriter.emitCode("%[import '%L';\n%]", augment.source)
                  }
                }
              }

         }

      codeWriter.emit("\n")
    }

    members.filterNot { it is ModuleSpec }.forEach { member ->
      codeWriter.emit("\n")
      when (member) {
        is ModuleSpec -> member.emit(codeWriter)
        is InterfaceSpec -> member.emit(codeWriter)
        is ClassSpec -> member.emit(codeWriter)
        is EnumSpec -> member.emit(codeWriter)
        is FunctionSpec -> member.emit(codeWriter, null, setOf(
           Modifier.PUBLIC))
        is PropertySpec -> member.emit(codeWriter, setOf(
           Modifier.PUBLIC), asStatement = true)
        is TypeAliasSpec -> member.emit(codeWriter)
        else -> throw AssertionError()
      }
    }

    members.filterIsInstance<ModuleSpec>().forEach { member ->
      codeWriter.emit("\n")
      member.emit(codeWriter)
    }


  }

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (other == null) return false
    if (javaClass != other.javaClass) return false
    return toString() == other.toString()
  }

  override fun hashCode() = toString().hashCode()

  override fun toString() = buildString { writeTo(this) }

  fun isEmpty(): Boolean {
    return members.isEmpty()
  }

  fun isNotEmpty(): Boolean {
    return !isEmpty()
  }

  fun toBuilder(): Builder {
    val builder = Builder(path)
    builder.comment.add(comment)
    builder.members.addAll(this.members)
    builder.indent = indent
    return builder
  }

  class Builder internal constructor(
     internal val path: Path
  ) {
    internal val comment = CodeBlock.builder()
    internal var indent = "  "
    internal val members = mutableListOf<Any>()

    private fun checkMemberModifiers(modifiers: Set<Modifier>) {
      requireNoneOf(
         modifiers,
         Modifier.PUBLIC,
         Modifier.PROTECTED,
         Modifier.PRIVATE,
         Modifier.READONLY,
         Modifier.GET,
         Modifier.SET,
         Modifier.STATIC,
         Modifier.CONST,
         Modifier.LET,
         Modifier.VAR
      )
    }

    fun addComment(format: String, vararg args: Any) = apply {
      comment.add(format, *args)
    }

    fun addModule(moduleSpec: ModuleSpec) = apply {
      members += moduleSpec
    }

    fun addClass(classSpec: ClassSpec) = apply {
      checkMemberModifiers(classSpec.modifiers)
      members += classSpec
    }

    fun addInterface(ifaceSpec: InterfaceSpec) = apply {
      checkMemberModifiers(ifaceSpec.modifiers)
      members += ifaceSpec
    }

    fun addEnum(enumSpec: EnumSpec) = apply {
      checkMemberModifiers(enumSpec.modifiers)
      members += enumSpec
    }

    fun addFunction(functionSpec: FunctionSpec) = apply {
      require(!functionSpec.isConstructor) { "cannot add ${functionSpec.name} to file $path" }
      require(functionSpec.decorators.isEmpty()) { "decorators on module functions are not allowed" }
      checkMemberModifiers(functionSpec.modifiers)
      members += functionSpec
    }

    fun addProperty(propertySpec: PropertySpec) = apply {
      requireExactlyOneOf(propertySpec.modifiers, Modifier.CONST,
                                                    Modifier.LET,
                                                    Modifier.VAR)
      require(propertySpec.decorators.isEmpty()) { "decorators on file properties are not allowed" }
      checkMemberModifiers(propertySpec.modifiers)
      members += propertySpec
    }

    fun addTypeAlias(typeAliasSpec: TypeAliasSpec) = apply {
      members += typeAliasSpec
    }

    fun indent(indent: String) = apply {
      this.indent = indent
    }

    fun isEmpty(): Boolean {
      return members.isEmpty()
    }

    fun isNotEmpty(): Boolean {
      return !isEmpty()
    }

    fun build() = FileSpec(this)
  }

  companion object {

    @JvmStatic
    fun builder(fileName: String, directory: Path) = Builder(
       directory.resolve(fileName))

    @JvmStatic
    fun builder(filePath: String) = Builder(Paths.get(filePath))

    @JvmStatic
    fun builder(filePath: Path) = Builder(filePath)

    @JvmStatic
    fun builder(module: ModuleSpec): Builder {
      val file = builder(module.name)
      file.members.addAll(module.members.toMutableList())
      return file
    }

  }

}

File: src/FileSpec.ts
import { imm, Imm } from 'ts-imm';
import { ClassSpec } from './ClassSpec';
import { CodeBlock } from './CodeBlock';
import { CodeWriter } from './CodeWriter';
import { EnumSpec } from './EnumSpec';
import { FunctionSpec } from './FunctionSpec';
import { InterfaceSpec } from './InterfaceSpec';
import { Modifier } from './Modifier';
import { PropertySpec } from './PropertySpec';
import { StringBuffer } from './StringBuffer';
import { SymbolSpec, SymbolSpecs } from './SymbolSpecs';
import { TypeAliasSpec } from './TypeAliasSpec';
import { TypeName, TypeNames } from './TypeNames';
import { filterInstances } from './utils';

/**
 * A TypeScript file containing top level objects like classes, objects, functions, properties, and type
 * aliases.
 *
 * Items are output in the following order:
 * - Comment
 * - Imports
 * - Members
 */
export class FileSpec extends Imm<FileSpec> {
  /**
   * Creates a file to contain generated output.
   *
   * `file` should be the module name, e.g. for build/foo.ts, use file = foo.
   */
  public static create(file: string): FileSpec {
    return new FileSpec({
      path: file,
      comment: CodeBlock.empty(),
      members: [],
      indentField: '  ',
    });
  }

  @imm
  public readonly path!: string;
  @imm
  public readonly comment!: CodeBlock;
  @imm
  public readonly members!: any[];
  @imm
  public readonly indentField!: string;

  public exportType(typeName: string): TypeName | undefined {
    const typeNameParts = typeName.split('.');
    const en = this.exportNamed(typeNameParts[0]);
    return en === undefined ? en : TypeNames.anyType(typeName, en);
  }

  public exportNamed(symbolName: string): SymbolSpec | undefined {
    const first = this.members
      .map(it => {
        if (it instanceof ClassSpec) {
          return it.name;
        } else if (it instanceof InterfaceSpec) {
          return it.name;
        } else if (it instanceof EnumSpec) {
          return it.name;
        } else if (it instanceof FunctionSpec) {
          return it.name;
        } else if (it instanceof PropertySpec) {
          return it.name;
        } else if (it instanceof TypeAliasSpec) {
          return it.name;
        } else {
          throw new Error('unrecognized member type');
        }
      })
      .filter(name => name === symbolName)[0];
    return first ? SymbolSpecs.importsName(symbolName, '!' + this.path) : undefined;
  }

  public exportAll(localName: string): SymbolSpec {
    return SymbolSpecs.importsAll(localName, '!' + this.path);
  }

  public emit(out: StringBuffer) {
    // First pass: emit the entire class, just to collect the types we'll need to import.
    const importsCollector = new CodeWriter(new StringBuffer(), this.indentField);
    this.emitToWriter(importsCollector);
    const requiredImports = importsCollector.requiredImports();
    // Second pass: write the code, taking advantage of the imports.
    const codeWriter = new CodeWriter(out, this.indentField, new Set(requiredImports));
    this.emitToWriter(codeWriter);
  }

  public isEmpty(): boolean {
    return this.members.length === 0;
  }

  public isNotEmpty(): boolean {
    return !this.isEmpty();
  }

  public addComment(format: string, ...args: any[]): this {
    return this.copy({
      comment: this.comment.add(format, ...args),
    });
  }

  public addClass(classSpec: ClassSpec): this {
    checkMemberModifiers(classSpec.modifiers);
    return this.copy({
      members: [...this.members, classSpec],
    });
  }

  public addInterface(ifaceSpec: InterfaceSpec): this {
    checkMemberModifiers(ifaceSpec.modifiers);
    return this.copy({
      members: [...this.members, ifaceSpec],
    });
  }

  public addEnum(enumSpec: EnumSpec): this {
    checkMemberModifiers(enumSpec.modifiers);
    return this.copy({
      members: [...this.members, enumSpec],
    });
  }

  public addFunction(functionSpec: FunctionSpec): this {
    // require(!functionSpec.isConstructor) { "cannot add ${functionSpec.name} to file $path" }
    // require(functionSpec.decorators.isEmpty()) { "decorators on module functions are not allowed" }
    checkMemberModifiers(functionSpec.modifiers);
    return this.copy({
      members: [...this.members, functionSpec],
    });
  }

  public addProperty(propertySpec: PropertySpec): this {
    // requireExactlyOneOf(propertySpec.modifiers, Modifier.CONST, Modifier.LET, Modifier.VAR)
    // require(propertySpec.decorators.isEmpty()) { "decorators on file properties are not allowed" }
    checkMemberModifiers(propertySpec.modifiers);
    return this.copy({
      members: [...this.members, propertySpec],
    });
  }

  public addTypeAlias(typeAliasSpec: TypeAliasSpec): this {
    return this.copy({
      members: [...this.members, typeAliasSpec],
    });
  }

  public addCode(codeBlock: CodeBlock): this {
    return this.copy({
      members: [...this.members, codeBlock],
    });
  }

  public indent(indent: string): this {
    return this.copy({
      indentField: indent,
    });
  }

  public toString(): string {
    const out = new StringBuffer();
    this.emit(out);
    return out.toString();
  }

  private emitToWriter(codeWriter: CodeWriter) {
    if (this.comment.isNotEmpty()) {
      codeWriter.emitComment(this.comment);
    }

    codeWriter.emitImports(this.path.replace(/\.tsx?$/, ''));

    this.members
      .filter(it => !(it instanceof CodeBlock))
      .forEach(member => {
        codeWriter.emit('\n');
        if (member instanceof InterfaceSpec) {
          member.emit(codeWriter);
        } else if (member instanceof ClassSpec) {
          member.emit(codeWriter);
        } else if (member instanceof EnumSpec) {
          member.emit(codeWriter);
        } else if (member instanceof FunctionSpec) {
          member.emit(codeWriter, [Modifier.PUBLIC]);
        } else if (member instanceof PropertySpec) {
          member.emit(codeWriter, [Modifier.PUBLIC], true);
        } else if (member instanceof TypeAliasSpec) {
          member.emit(codeWriter);
        } else if (member instanceof CodeBlock) {
          codeWriter.emitCodeBlock(member);
        } else {
          throw new Error('unhandled');
        }
      });

    filterInstances(this.members, CodeBlock).forEach(member => {
      codeWriter.emit('\n');
      codeWriter.emitCodeBlock(member);
    });
  }
}

// /** Writes this to `directory` as UTF-8 using the standard directory structure.  */
// public writeTo(directory: Path) {
//   require(Files.notExists(directory) || Files.isDirectory(directory)) {
//     "path $directory exists but is not a directory."
//   }
//   const outputPath = directory.resolve("$path.ts")
//   OutputStreamWriter(Files.newOutputStream(outputPath), UTF_8).use { writer -> writeTo(writer) }
// }
//
// /** Writes this to `directory` as UTF-8 using the standard directory structure.  */
// public writeTo(directory: File) = writeTo(directory.toPath())

function checkMemberModifiers(modifiers: Modifier[]): void {
  // requireNoneOf(
  //   modifiers,
  //   Modifier.PUBLIC,
  //   Modifier.PROTECTED,
  //   Modifier.PRIVATE,
  //   Modifier.READONLY,
  //   Modifier.GET,
  //   Modifier.SET,
  //   Modifier.STATIC,
  //   Modifier.CONST,
  //   Modifier.LET,
  //   Modifier.VAR
  // )
}
